docs/designers-developers/developers/block-api/block-context.md

```diff
@@ -0,0 +1,70 @@
+# Block Context
+
+Block context is a feature which enables ancestor blocks to provide values which can be consumed by descendent blocks within its own hierarchy. Those descendent blocks can inherit this value without resorting to hard-coded values, or without an explicit awareness of the block which provides those values.
+
+This is especially useful in full-site editing where, for example, the contents of a block may depend on the context of the post in which it is relevant. A blogroll template may show excerpts of many different posts. Using block context, there can still be one single "Post Excerpt" block which displays the contents of the post based on an inherited post ID.
+
+If you are familiar with [React Context](https://reactjs.org/docs/context.html), the idea is very similar. In fact, the client-side block editor implementation of block context is a very simple application of React Context. Block context is also supported in server-side `render_callback` implementations, demonstrated in the examples below.
+
+## Defining Block Context
+
+Block context is defined in the registered settings of a block. A block can provide a context value, or consume a value it seeks to inherit.
+
+### Providing Block Context
+
+A block can provide a context value by assigning a `providesContext` property in its registered settings. This should be assigned as an array of the attribute names for attribute values to assign as context. Currently, block context only supports values derived from the block's own attributes. This could be enhanced in the future to support additional sources of context values.
+
+_block.json_
+
+```json
+{
+	"name": "core/post",
+	"attributes": {
+		"postId": {
+			"type": "number"
+		}
+	},
+	"providesContext": [ "postId" ]
+}
+```
+
+### Consuming Block Context
+
+A block can inherit a context value from an ancestor provider by assigning a `context` property in its registered settings. This should be assigned as an array of the context names the block seeks to inherit.
+
+_block.json_
+
+```json
+{
+	"name": "core/post-excerpt",
+	"context": [ "postId" ]
+}
+```
+
+## Using Block Context
+
+Once a block has defined the context it seeks to inherit, this can be accessed in the implementation of `edit` (JavaScript) and `render_callback` (PHP). It is provided as an object (JavaScript) or associative array (PHP) of the context values which have been defined for the block. Note that even if there is an ancestor which provides a context value, the value will only be made available if the block explicitly defines a desire to inherit that value.
```

Comment:
> Note that even if there is an ancestor which provides a context value, the value will only be made available if the block explicitly defines a desire to inherit that value.

Why? I saw that @youknowriad asked this in https://github.com/WordPress/gutenberg/issues/19685#issuecomment-601350078 but couldn't find an answer.

Could the key/values in `context` not just be a grab bag of whatever values happen to have been provided by ancestor blocks?

Reply:
I like it being explicit as it mirrors how it is implemented in React. You need to explicitly use the context consumer to get value set by providers.

Reply:
I also like how it declares a type of dependency on a value for a block. I think it'll make it easier to match up providers to consumers.

Reply:
> > Note that even if there is an ancestor which provides a context value, the value will only be made available if the block explicitly defines a desire to inherit that value.
> 
> Why? I saw that @youknowriad asked this in [#19685 (comment)](https://github.com/WordPress/gutenberg/issues/19685#issuecomment-601350078) but couldn't find an answer.
> 
> Could the key/values in `context` not just be a grab bag of whatever values happen to have been provided by ancestor blocks?

The quoted text was intended to be most relevant for what I'd mentioned in the original comment about _always at least providing an empty object_. But it's a fair question to consider whether a descendent even need to declare its `context` desires explicitly, vs. having access to whatever values happen to exist in the current context. To this, I think @gziolo and @talldan raised good points for why it ought to be done this way. Just having the awareness of a block's context needs could open up some potential future enhancements. For example, for what you'd mentioned in https://github.com/WordPress/gutenberg/pull/21467#issuecomment-610705719 for what should happen in the case that a block _assumes_ `postId` and `postType` to exist, we could have the option to handle this (warnings, graceful fallback) if we know a block has context desires which are not satisfied.